include/mesh/mesh_base.h

```diff
@@ -1945,6 +1968,29 @@ class MeshBase : public ParallelObject
    */
   std::set<subdomain_id_type> _mesh_subdomains;
 
+  /**
+   * Map from "element set code" to list of set ids to which that element
+   * belongs (and vice-versa). Remarks:
+   * 1.) The elemset code is a dof_id_type because (if used) it is
+   * stored as an extra_integer (named "elemset_code") on all elements,
+   * and extra_integers are of type dof_id_type. Elements which do not
+   * belong to any set should be assigned an elemset code of DofObject::invalid_id.
+   * 2.) Element sets can be thought of as a generalization of the concept
+   * of a subdomain. Subdomains have the following restrictions:
+   *   a.) A given element can only belong to a single subdomain
+   *   b.) When using Exodus file input/output, subdomains are (unfortunately)
+   *   tied to the concept of exodus element blocks, which consist of a single
+   *   geometric element type, somewhat limiting their generality.
+   * 3.) The user is responsible for filling in the values of this map
+   * in a consistent manner, unless the elemsets are read in from an
+   * Exodus file, in which case the elemset codes will be set up
+   * automatically. The codes can basically be chosen arbitrarily,
+   * with the one requirement that elements which belong to no sets
+   * should have a set code of DofObject::invalid_id.
+   */
+  std::map<dof_id_type, std::set<elemset_id_type>> _elemset_codes;
+  std::map<std::set<elemset_id_type>, dof_id_type> _elemset_codes_inverse_map;
```

Comment:
Since we're keeping these in sync invariably and encapsulating them properly, could we avoid duplicating the set data somehow?  Make the values in _elemset_codes be just pointers to the keys in _elemset_codes_inverse_map, perhaps?  That wouldn't work with `unordered_map` (rehashing invalidates iterators) but it ought to be well-defined behavior with `map`.

Reply:
I would like to look be able to quickly look up in both directions though. Maybe that is possible in the approach you described, not sure. There is a bi-directional map in boost I think, which is basically what I'm going for here with two maps.

Reply:
It's possible with the approach I was suggesting (unless we change to `unordered_map`, which is worth considering!), yes.  All that would change is that everywhere you would have read `_elemset_codes[foo]` you instead use `*_elemset_codes[foo]`.

Reply:
OK, so how do you add an entry to a map where the key is a pointer to the value? As mentioned in the other thread, currently I have
```
void MeshBase::add_elemset_code(dof_id_type code, const std::set<elemset_id_type> & id_set)
{
  _elemset_codes.emplace(code, id_set);
  _elemset_codes_inverse_map.emplace(id_set, code);
}
```
are you proposing something like
```
void MeshBase::add_elemset_code(const std::set<elemset_id_type> & id_set)
{
  _elemset_codes.emplace(&id_set, id_set);
}
```
?

Reply:
Oh, wait I think I misread your suggestion. 

> Make the values in _elemset_codes be just pointers to the keys in _elemset_codes_inverse_map

So there are still two maps?

Reply:
Yup, still two maps.  You still `map::emplace` into `_elemset_codes_inverse_map`, but you catch the returned iterator and take the address of the newly emplaced set to use in `_elemset_codes`.

Reply:
I implemented the approach of storing a pointer-to-set rather than two copies of the set. I personally don't expect these sets to be "big" or for there to be too many of them in practice... but in case there is I think we have the most efficient storage approach now.